pyccl/halos/hmfunc.py

```diff
@@ -0,0 +1,647 @@
+        of this object. This function is called at the very end of the
+        constructor call.
+
+        Args:
+            cosmo (:obj:`Cosmology`): A Cosmology object.
+        """
+        pass
+
+    def _check_mdef(self, mdef):
+        """ Return False if the input mass definition agrees with
+        the definitions for which this mass function parametrization
+        works. True otherwise. This function gets called at the
+        start of the constructor call.
+
+        Args:
+            mdef (:obj:`MassDef`): a mass definition object.
+
+        Returns:
+            bool: True if the mass definition is not compatible with
+                this mass function parametrization. False otherwise.
+        """
+        return False
+
+    def _get_consistent_mass(self, cosmo, M, a, mdef_other):
+        """ Transform a halo mass with a given mass definition into
+        the corresponding mass definition that was used to initialize
+        this object.
+
+        Args:
+            cosmo (:obj:`Cosmology`): A Cosmology object.
+            M (float or array_like): halo mass in units of M_sun.
+            a (float): scale factor.
+            mdef_other (:obj:`MassDef`): a mass definition object.
+
+        Returns:
+            float or array_like: mass according to this object's
+            mass definition.
+        """
+        if mdef_other is not None:
+            M_use = mdef_other.translate_mass(cosmo, M, a, self.mdef)
+        else:
+            M_use = M
+        return np.log10(M_use)
+
+    def get_mass_function(self, cosmo, M, a, mdef_other=None):
+        """ Returns the mass function for input parameters.
+
+        Args:
+            cosmo (:obj:`Cosmology`): A Cosmology object.
+            M (float or array_like): halo mass in units of M_sun.
+            a (float): scale factor.
+            mdef_other (:obj:`MassDef`): the mass definition object
+                that defines M.
+
+        Returns:
+            float or array_like: mass function d(n)/d(log10(M))
+                in units of Mpc^-3.
+        """
+        M_use = np.atleast_1d(M)
+        logM = self._get_consistent_mass(cosmo, M_use,
+                                         a, mdef_other)
+
+        # sigma(M)
+        status = 0
+        sigM, status = lib.sigM_vec(cosmo.cosmo, a, logM,
+                                    len(logM), status)
+        check(status)
+        # dlogsigma(M)/dlog10(M)
+        dlns_dlogM, status = lib.dlnsigM_dlogM_vec(cosmo.cosmo, logM,
+                                                   len(logM), status)
+        check(status)
+
+        rho = (lib.cvar.constants.RHO_CRITICAL *
+               cosmo['Omega_m'] * cosmo['h']**2)
+        f = self.get_fsigma(cosmo, sigM, a, 2.302585092994046 * logM)
+        mf = f * rho * dlns_dlogM / M_use
+
+        if np.ndim(M) == 0:
+            mf = mf[0]
+        return mf
+
+    def get_fsigma(self, cosmo, sigM, a, lnM):
```

Comment:
Same comment on API and functional forms and emulators.

Reply:
Same as above.

Reply:
Same comment about the public API as above as well!

Reply:
Done as above